File: java/src/coz/Coz.java
/*
 * Copyright (c) 2015, Charlie Curtsinger and Emery Berger,
 *                     University of Massachusetts Amherst
 * This file is part of the Coz project. See LICENSE.md file at the top-level
 * directory of this distribution and at http://github.com/plasma-umass/coz.
 */

package coz;

/**
 * Progress points for the coz causal profiler.
 *
 * <p>Coz answers "if I optimize this line, will the program actually get faster?" Mark the points
 * where your program makes progress, then run it under the coz JVMTI agent:
 *
 * <pre>
 *   javac -g -d classes MyApp.java
 *   java -agentpath:/path/to/libcozjava.so=output=profile.jsonl,scope=com.example \
 *        -cp classes:coz.jar com.example.MyApp
 *   coz plot --text -i profile.jsonl
 * </pre>
 *
 * <p>When the agent is not loaded every method here is a no-op, so instrumented code runs
 * normally outside profiling.
 *
 * <p>All methods are safe to call from any thread.
 */
public final class Coz {

    /**
     * Whether the coz agent is attached.
     *
     * <p>Deliberately in a holder class rather than a static field of {@code Coz}. The agent binds
     * the natives below with RegisterNatives during VM init, and to find this class it calls JNI
     * FindClass -- which *initializes* it. A {@code static final boolean ENABLED = probe()} on
     * {@code Coz} would therefore run before the natives were bound, see UnsatisfiedLinkError, and
     * latch "not profiling" for the life of the VM.
     *
     * <p>The holder is only initialized on first use from application code, which is necessarily
     * after VM init, so by then the natives are bound.
     */
    private static final class Enabled {
        static final boolean VALUE = probe();

        private static boolean probe() {
            try {
                return available0();
            } catch (UnsatisfiedLinkError e) {
                return false;
            }
        }
    }

    private Coz() {}

    /** Whether this JVM is running under the coz agent. */
    public static boolean isAvailable() {
        return Enabled.VALUE;
    }

    /**
     * Records a visit to the named throughput progress point: "I wish this happened more often."
     * The equivalent of {@code COZ_PROGRESS_NAMED}.
     */
    public static void progress(String name) {
        if (Enabled.VALUE) {
            progress0(name);
        }
    }

    /**
     * Records a visit to a throughput progress point named after the calling class, method and
     * line. The equivalent of {@code COZ_PROGRESS}.
     *
     * <p>Determining the call site walks one stack frame. On a hot path, prefer
     * {@link #progress(String)}.
     */
    public static void progress() {
        if (Enabled.VALUE) {
            progress0(callSite());
        }
    }

    /**
     * Marks the start of a latency-profiled operation: "I wish this finished sooner." The
     * equivalent of {@code COZ_BEGIN}. Must be paired with {@link #end(String)}.
     */
    public static void begin(String name) {
        if (Enabled.VALUE) {
            begin0(name);
        }
    }

    /**
     * Marks the completion of a latency-profiled operation, the equivalent of {@code COZ_END}.
     * Must be paired with {@link #begin(String)}.
     */
    public static void end(String name) {
        if (Enabled.VALUE) {
            end0(name);
        }
    }

    /**
     * Runs {@code body} bracketed by {@link #begin(String)} and {@link #end(String)}, so the end
     * counter fires even if the body throws.
     *
     * <pre>
     *   Coz.scope("request", () -&gt; handle(request));
     * </pre>
     */
    public static void scope(String name, Runnable body) {
        begin(name);
        try {
            body.run();
        } finally {
            end(name);
        }
    }

    /** Renders the caller of {@code progress()} as "Class.method:line". */
    private static String callSite() {
        return StackWalker.getInstance()
                .walk(frames -> frames
                        .skip(2) // callSite, progress
                        .findFirst()
                        .map(f -> f.getClassName() + "." + f.getMethodName() + ":" + f.getLineNumber())
                        .orElse("unknown"));
    }

    // Bound by the agent's RegisterNatives at VM init.
    private static native boolean available0();

    private static native void progress0(String name);

    private static native void begin0(String name);

    private static native void end0(String name);
}

File: swift/Sources/Coz/Coz.swift
/*
 * Copyright (c) 2015, Charlie Curtsinger and Emery Berger,
 *                     University of Massachusetts Amherst
 * This file is part of the Coz project. See LICENSE.md file at the top-level
 * directory of this distribution and at http://github.com/plasma-umass/coz.
 */

import CCoz

/// Swift support for the coz causal profiler.
///
/// Coz answers "if I optimize this line, will the program actually get faster?"
/// Mark the points where your program makes progress, build with debug info,
/// and run the binary under `coz run`:
///
/// ```
/// swift build -c release -Xswiftc -g
/// coz run --- .build/release/MyApp
/// ```
///
/// When the program is not running under `coz run` the libcoz runtime is absent
/// and every entry point here is a no-op.
public enum Coz {
    /// Whether the coz runtime was found, i.e. whether this process is running
    /// under `coz run`.
    public static var isAvailable: Bool {
        coz_shim_available() != 0
    }

    /// A handle to one of libcoz's progress-point counters.
    ///
    /// The convenience functions below look a counter up by name on each call,
    /// which takes a lock. On a hot path, hold a `Counter` instead:
    ///
    /// ```swift
    /// let requests = Coz.Counter(throughput: "requests")
    /// for request in requests { handle(request); requests.increment() }
    /// ```
    ///
    /// A `Counter` obtained while not profiling is inert.
    // @unchecked because a raw pointer is not Sendable on its face. libcoz owns
    // this memory, allocates it once, and keeps it alive for the life of the
    // process; the only mutation is the relaxed atomic increment inside
    // coz_shim_hit. So sharing a Counter across threads is safe.
    public struct Counter: @unchecked Sendable, Equatable {
        private let counter: UnsafeMutableRawPointer?

        private init(type: Int32, name: String) {
            let ptr = name.withCString { coz_shim_cached_counter(type, $0) }
            self.counter = UnsafeMutableRawPointer(ptr)
        }

        /// A throughput counter, the equivalent of `COZ_PROGRESS_NAMED`.
        public init(throughput name: String) {
            self.init(type: Int32(COZ_SHIM_THROUGHPUT), name: name)
        }

        /// The opening counter of a latency pair, the equivalent of `COZ_BEGIN`.
        public init(begin name: String) {
            self.init(type: Int32(COZ_SHIM_BEGIN), name: name)
        }

        /// The closing counter of a latency pair, the equivalent of `COZ_END`.
        public init(end name: String) {
            self.init(type: Int32(COZ_SHIM_END), name: name)
        }

        /// Records one visit to this progress point. Safe to call concurrently.
        public func increment() {
            coz_shim_hit(counter?.assumingMemoryBound(to: coz_counter_t.self))
        }
    }

    /// Records a visit to a throughput progress point, the equivalent of
    /// `COZ_PROGRESS`.
    ///
    /// The point is named after the call site unless `name` is given.
    public static func progress(
        _ name: String? = nil,
        file: String = #fileID,
        line: Int = #line
    ) {
        Counter(throughput: name ?? "\(file):\(line)").increment()
    }

    /// Marks the start of a latency-profiled operation, the equivalent of
    /// `COZ_BEGIN`. Must be paired with `end(_:)` of the same name.
    public static func begin(_ name: String) {
        Counter(begin: name).increment()
    }

    /// Marks the completion of a latency-profiled operation, the equivalent of
    /// `COZ_END`. Must be paired with `begin(_:)` of the same name.
    public static func end(_ name: String) {
        Counter(end: name).increment()
    }

    /// Runs `body` bracketed by `begin(name)` and `end(name)`.
    ///
    /// The end counter fires even if `body` throws.
    ///
    /// ```swift
    /// try Coz.scope("request") { try handle(request) }
    /// ```
    public static func scope<T>(_ name: String, _ body: () throws -> T) rethrows -> T {
        begin(name)
        defer { end(name) }
        return try body()
    }

    // MARK: - Custom synchronization
    //
    // Coz interposes pthread mutexes, condition variables and joins, so it knows
    // when a thread blocks and does not charge it for delays it could not have
    // paid. It cannot see primitives it does not interpose -- raw futexes, Mach
    // semaphores, `DispatchSemaphore` -- and a thread blocked on one of those
    // will be charged anyway, skewing every measurement it takes part in.
    //
    // These mirror COZ_PRE_BLOCK / COZ_POST_BLOCK / COZ_CATCH_UP in coz.h and
    // let you tell coz about such a primitive:
    //
    //     Coz.preBlock()
    //     semaphore.wait()
    //     Coz.postBlock(skipDelays: true)
    //
    //     Coz.catchUp()          // before a signal that may unblock someone
    //     semaphore.signal()

    /// Call immediately before blocking on a primitive coz does not interpose.
    public static func preBlock() {
        coz_shim_pre_block()
    }

    /// Call immediately after such a block returns.
    ///
    /// Pass `skipDelays: true` when the wake-up came from another thread (the
    /// usual case: a semaphore was signalled, a lock was handed over), and
    /// `false` when it may have been spurious or a timeout.
    public static func postBlock(skipDelays: Bool) {
        coz_shim_post_block(skipDelays ? 1 : 0)
    }

    /// Pay any outstanding virtual delay before doing something that might
    /// unblock another thread.
    public static func catchUp() {
        coz_shim_catch_up()
    }
}
